File: week-4/odd-even-linked-list/odd-even-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var oddEvenList = function(head) {
    let dummy1 = new ListNode(-1);
    let dummy2 = new ListNode(-1);
    let odd = dummy1;
    let even = dummy2;
    
    while (head) {
        odd.next = head;
        even.next = head.next;
        
        head = head.next ? head.next.next : null;
        odd = odd.next;
        even = even.next;
    }
    
    odd.next = dummy2.next;
    return dummy1.next;
    
};

File: week-4/odd-even-linked-list/odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        dummy1 = ListNode(-1)
        dummy2 = ListNode(-1)
        odd = dummy1
        even = dummy2
        
        while head:
            odd.next = head
            even.next = head.next
            
            odd = odd.next
            even = even.next
            head = head.next.next if head.next else None
            
        odd.next = dummy2.next
        return dummy1.next
            
            
